contracts/libraries/PermissionsLib.sol

```diff
@@ -22,30 +22,38 @@ pragma solidity 0.4.18;
 library PermissionsLib {
     struct Permissions {
         mapping (address => bool) authorized;
+        mapping (address => uint) agentToIndex;
         address[] authorizedAgents;
     }
 
     function authorize(Permissions storage self, address agent)
         internal
     {
-        if (!self.authorized[agent]) {
-            self.authorized[agent] = true;
-            self.authorizedAgents.push(agent);
-        }
+        require(isNotAuthorized(self, agent));
+
+        self.authorized[agent] = true;
+        self.authorizedAgents.push(agent);
+        self.agentToIndex[agent] = self.authorizedAgents.length - 1;
     }
 
     function revokeAuthorization(Permissions storage self, address agent)
         internal
     {
+        /* We only want to do work in the case where the agent whose
+        authorization is being revoked had authorization permissions in the
+        first place */
+        require(isAuthorized(self, agent));
+
+        uint indexOfAgentToRevoke = self.agentToIndex[agent];
+        uint indexOfAgentToMove = self.authorizedAgents.length - 1;
+        address agentToMove = self.authorizedAgents[indexOfAgentToMove];
+
         delete self.authorized[agent];
-        for (uint i = 0; i < self.authorizedAgents.length; i++) {
-            if (self.authorizedAgents[i] == agent) {
-                self.authorizedAgents[i] =
-                    self.authorizedAgents[self.authorizedAgents.length - 1];
-                self.authorizedAgents.length -= 1;
-                break;
-            }
-        }
+        self.authorizedAgents[indexOfAgentToRevoke] =
+            self.authorizedAgents[indexOfAgentToMove];
+        self.authorizedAgents.length -= 1;
+        self.agentToIndex[agentToMove] = indexOfAgentToRevoke;
+        delete self.agentToIndex[agent];
```

Comment:
Let's kill two birds with one stone and add the [edge case fix](https://app.asana.com/0/555346636604435/555346636604425/f) referred to in Zeppelin's review.  Should be a one liner.

Reply:
I'd vote for keeping this PR isolated to this one change. Easy enough to create another branch for that change. Will work on that now.

Reply:
SGTM 👍